Axes.inset_axes: enable Axes subclass creation

rcomer opened the pull request:
## PR Summary

Closes #22630 

Modify `Axes.inset_axes` so it can accept the `polar`, `projection` and `axes_class` keywords, and therefore return subclasses of `Axes`.  My target use-case is with Cartopy, so we can now adapt [this gallery example](https://scitools.org.uk/cartopy/docs/latest/gallery/scalar_data/waves.html#sphx-glr-gallery-scalar-data-waves-py) to do something like

```python
import cartopy.crs as ccrs
import matplotlib.pyplot as plt
import numpy as np

def sample_data(shape=(73, 145)):
    """Return ``lons``, ``lats`` and ``data`` of some fake data."""
    nlats, nlons = shape
    lats = np.linspace(-np.pi / 2, np.pi / 2, nlats)
    lons = np.linspace(0, 2 * np.pi, nlons)
    lons, lats = np.meshgrid(lons, lats)
    wave = 0.75 * (np.sin(2 * lats) ** 8) * np.cos(4 * lons)
    mean = 0.5 * np.cos(2 * lats) * ((np.sin(2 * lats)) ** 2 + 2)

    lats = np.rad2deg(lats)
    lons = np.rad2deg(lons)
    data = wave + mean

    return lons, lats, data


lons, lats, data = sample_data()

ax = plt.subplot(111, projection=ccrs.PlateCarree())
ax.contourf(lons, lats, data)
ax.coastlines()

axins = ax.inset_axes([0.5, 0.03, 0.47, 0.47], projection=ccrs.PlateCarree())
axins.contourf(lons, lats, data)
axins.set_xlim(-10, 5)
axins.set_ylim(50, 60)
axins.coastlines()

box, lines = ax.indicate_inset_zoom(axins, edgecolor="black")

plt.show()
```

![cartopy_inset](https://user-images.githubusercontent.com/10599679/156900775-075331ba-4861-4396-a2ff-bc9b920a4f7d.png)


## PR Checklist

<!-- Please mark any checkboxes that do not apply to this PR as [N/A]. -->
**Tests and Styling**
- [x] Has pytest style unit tests (and `pytest` passes). _**I'm unsure what extra tests should be added for this.  Advice welcome!**_
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [x] New features are documented, with examples if plot related.
- [x] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [N/A] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Do not create the PR out of main, but out of a separate branch.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

rcomer commented:
Looks like I have broken [one test](https://github.com/matplotlib/matplotlib/blob/808dbe4a709d286b6dd51c06f111259deaf3f106/lib/matplotlib/tests/test_axes.py#L6691) consistently across platforms.  I will have to come back to this another day.

jklymak commented:
Inset axes are explicitly children of their parent axes so please don't also add the figure as a parent.  I'm not quite sure what the best way to get a projection into an inset_axes is...

greglucas commented:
This seems like a useful thing to have available in some form. Might be worth having a look at https://github.com/matplotlib/matplotlib/pull/22060 and whether https://github.com/matplotlib/matplotview would work with GeoAxes too...

Ahh, the axes being a child of the figure or not would actually explain why mpl_toolkit inset_axes warn with `tight_layout()` (because they are added!), but standard inset_axes don't. (xref: #22611)
https://github.com/matplotlib/matplotlib/blob/e75b5291108bf0f4d15785612c7a7899d07ea9a2/lib/mpl_toolkits/axes_grid1/inset_locator.py#L294

The easiest way might be to allow `axes_class` to be passed in to the `inset_axes()` method, similar to mpl_toolkits and figure's add_axes methods.
https://github.com/matplotlib/matplotlib/blob/e75b5291108bf0f4d15785612c7a7899d07ea9a2/lib/mpl_toolkits/axes_grid1/inset_locator.py#L380-L385

However, there might be some more surgery needed outside of AxesGrid to support the projection or other keyword args.
https://github.com/matplotlib/matplotlib/blob/09b17b128440e18cec4a0958dd2dc5b86c9bd146/lib/matplotlib/figure.py#L535-L538

rcomer commented:
Thanks for the feedback.  I did have a hunch that it couldn't really be this simple!

> Inset axes are explicitly children of their parent axes so please don't also add the figure as a parent.

I'm afraid I don't have the background knowledge to understand this.  Is there somewhere I can read up on the matplotlib parent/child relationships and why they matter?  I guess the comment about `tight_layout` gives a clue.

> matplotlibview

Looks useful - thanks!  I will have a play with that at some point.

> The easiest way might be to allow axes_class to be passed in to the inset_axes() method, similar to mpl_toolkits and figure's add_axes methods.

I'm not so keen on that API as it seems inconsistent with how the user is usually encouraged to create their Axes instances.  But I guess it was done in those other functions for a reason.

Anyway it seems that whatever the answer is, it's not this.  So for now I will just close this.

jklymak commented:
@rcomer, maybe convert this to an issue?   I think it is reasonable to ask that inset_axes accepts projections, I'm just not sure off the top of my head how to do it (I have limited Matplotlib (and internet) bandwidth at the moment.

rcomer commented:
> maybe convert this to an issue?

Issue now at #22630.  Thanks!

jklymak commented:
@rcomer, did this not work?   It looks like something similar _should_ work, and it is certainly reasonable functionality...

rcomer commented:
@jklymak yes the CI all seemed fine and it works for my Cartopy example.  However at #22630 @mylasem said they were working on it, so I thought I would leave it to them for now.  I'd be happy to pick this up again at some point if @mylasem's version doesn't land for whatever reason.

jklymak commented:
In general, Matplotlib does not have a "reserve an issue" policy, and this PR predates @mylasem 's comment.  I think they should work within the context of this PR if at all possible.

rcomer commented:
OK, since @mylasem has not commented further and I had some time, I have:

* Added some basic tests to verify that the expected type is returned for our three methods of specifying it.
* Updated the `Axes.inset_axes` docstring with the keyword descriptions - this was a straight copy/paste from the `Figure.add_axes` docstring, and then added the word "inset".
* Updated the what's new entry (and the PR title and OP) to reflect that we're not going via `Figure.add_axes`.

rcomer commented on lib/matplotlib/tests/test_axes.py:
```diff
@@ -6614,6 +6614,30 @@ def test_zoom_inset():
 
 
+def test_inset_polar():
+    from matplotlib.projections.polar import PolarAxes
```
I imported the class within the test as I noticed another test in this module does that.  If it's not necessary, then I think we could parametrize these three tests into one.

tacaswell replied:
I do not see any reason to do this, I think pulling all of the imports to the top is OK.

rcomer replied:
I moved the imports to the top, but in the end couldn't convince myself it was worth parametrizing the two small tests that are still the same: since they're so small there isn't _much_ repetition going on and I suspect readability is better as they are than with a parametrized version.

jklymak commented on lib/matplotlib/tests/test_axes.py:
```diff
@@ -6614,6 +6614,30 @@ def test_zoom_inset():
+
+    _, ax = plt.subplots()
+    axins = ax.inset_axes([0.2, 0.2, 0.3, 0.3], polar=True)
+    assert isinstance(axins, PolarAxes)
```
This test seems good, but I do wonder if we actually need an image test here to make sure all the drawing gets through properly?  Otherwise this looks great.

rcomer replied:
@jklymak thanks.  I have to admit I've never used matplotlib's own projections before, so will have to do some reading and have a think.  If you happen to know of any specific examples I could adapt, that would be great.

rcomer replied:
@jklymak is something like [this](https://github.com/matplotlib/matplotlib/pull/22608/commits/452622505dd72a5ba6c6745e33772bcf63fc5ea5) sufficient?  From what I can tell, the zoom type plot won't work with polar, because the xlim/ylim don't correspond to a box.  But a user may have other reasons to want an inset plot.

tacaswell commented on lib/matplotlib/tests/test_axes.py:
```diff
@@ -6614,6 +6614,37 @@ def test_zoom_inset():
         axin1.get_position().get_points(), xx, rtol=1e-4)
 
 
+@image_comparison(['inset_polar.png'], remove_text=True)
```
```suggestion
@image_comparison(['inset_polar.png'], remove_text=True, style='mpl20')
```

But this will also require regenerating the test image.

rcomer replied:
done

tacaswell reviewed (approved):
I am :+1: with this going in as-is. 


Moving to the mpl20 style would be good as would refactoring the test, but neither is worth holding up the PR over.

rcomer commented:
Thanks @tacaswell, I am happy to follow those up tomorrow when I am back at my keyboard.  Should I also squash the commits down to one?

tacaswell commented:
> Should I also squash the commits down to one?

Yes please (and if you replace the image CI without squashing CI will fail).

QuLogic commented on doc/users/next_whats_new/inset_axes_improvements.rst:
```diff
@@ -0,0 +1,6 @@
+Axes.inset_axes Flexibility
+---------------------------
+
+`matplotlib.axes.Axes.inset_axes` now accepts the ``projection``, ``polar``
+and ``axes_class`` keywords, so that subclasses of `matplotlib.axes.Axes` may
```
Argument names should use italics.
```suggestion
`matplotlib.axes.Axes.inset_axes` now accepts the *projection*, *polar* and
*axes_class* keyword arguments, so that subclasses of `matplotlib.axes.Axes` may
```

QuLogic commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -326,13 +326,27 @@ def inset_axes(self, bounds, *, transform=None, zorder=5, **kwargs):
+            of a custom projection, see `~matplotlib.projections`. The default
+            None results in a 'rectilinear' projection.
+
+        polar : bool, default: False
+            If True, equivalent to projection='polar'.
```
We can't remove `polar` because it comes through the internal meothd, but I wonder if we want to not document it, since it duplicates `projection`? @timhoffm

I recall discussing it, and it couldn't really be dropped from normal since it was so widely used, but since this is new, we don't have to worry about existing users.

timhoffm replied:
Unfortunately, `polar` is a redundant API. OTOH `polar` is widely used and we have not even discouraged its use, I consider it proper first-class API. One can regard it as a shortcut.
Overall, it's there and it's valid to be used. Therefore I recommend to also document it.

rcomer replied:
I’m not sure where to go from here as there are differing opinions.  FWIW I think it’s intuitive that the various axes creation functions should take the same keywords.  So if I were used to using `polar` with the other functions I would expect to use it with this one too.  I’m not even sure whether I would check the docstring before trying it!

timhoffm replied:
Picking up my API authority hat, I decide that the present consistent way is the desired one. If we want to do something about `polar`, that'd need to be a separate discussion and a global change (but I advise against that).

tacaswell commented on lib/matplotlib/tests/test_axes.py:
```diff
@@ -6614,6 +6615,31 @@ def test_zoom_inset():
+@image_comparison(['inset_polar.png'], remove_text=True, style='mpl20')
+def test_inset_polar():
+    _, ax = plt.subplots()
+    axins = ax.inset_axes([0.5, 0.1, 0.45, 0.45], polar=True)
```
```suggestion
    axins = ax.inset_axes([0.5, 0.1, 0.45, 0.45], projection='polar')
```

given the above discussion?

rcomer replied:
Would that also mean that [this one](https://github.com/matplotlib/matplotlib/blob/80e672e0700fa8a268aed1bdefedbd0e493e91a7/lib/matplotlib/tests/test_axes.py#L6631-L6635) is redundant and should be removed?
